File: services/ai-service/src/rabbitmq/topology.py
"""Declares the exchanges and dead-letter topology this service depends on.
Every service that talks to RabbitMQ declares this same shape independently
(see docs/architecture/event-driven-model.md) instead of importing a shared
package, so each remains deployable on its own. Idempotent - whichever
service starts first "wins"."""
from __future__ import annotations

import pika

import config


def declare_topology(channel: "pika.adapters.blocking_connection.BlockingChannel") -> None:
    channel.exchange_declare(exchange=config.DOMAIN_EVENTS_EXCHANGE, exchange_type="topic", durable=True)
    channel.exchange_declare(exchange=config.ANALYTICS_EVENTS_EXCHANGE, exchange_type="topic", durable=True)
    channel.exchange_declare(exchange=config.COMMANDS_EXCHANGE, exchange_type="topic", durable=True)
    channel.exchange_declare(exchange=config.DOMAIN_EVENTS_DLX, exchange_type="topic", durable=True)
    channel.exchange_declare(exchange=config.COMMANDS_DLX, exchange_type="topic", durable=True)

    channel.queue_declare(queue=config.AI_DEAD_QUEUE, durable=True)
    channel.queue_bind(exchange=config.DOMAIN_EVENTS_DLX, queue=config.AI_DEAD_QUEUE, routing_key="#")

File: services/companies-service/src/rabbitmq/topology.ts
import type { Channel } from 'amqplib';

export const DOMAIN_EVENTS_EXCHANGE = 'domain.events';
export const ANALYTICS_EVENTS_EXCHANGE = 'analytics.events';
export const COMMANDS_EXCHANGE = 'commands';
export const DOMAIN_EVENTS_DLX = 'domain.events.dlx';
export const COMMANDS_DLX = 'commands.dlx';
export const COMPANIES_DEAD_QUEUE = 'companies.dead.q';

/**
 * Declares the exchanges and dead-letter topology this service depends on.
 * Every service that talks to RabbitMQ declares this same shape
 * independently (see docs/architecture/event-driven-model.md) instead of
 * importing a shared package, so each remains deployable on its own.
 * Idempotent - whichever service starts first "wins".
 */
export async function declareTopology(channel: Channel): Promise<void> {
  await channel.assertExchange(DOMAIN_EVENTS_EXCHANGE, 'topic', { durable: true });
  await channel.assertExchange(ANALYTICS_EVENTS_EXCHANGE, 'topic', { durable: true });
  await channel.assertExchange(COMMANDS_EXCHANGE, 'topic', { durable: true });
  await channel.assertExchange(DOMAIN_EVENTS_DLX, 'topic', { durable: true });
  await channel.assertExchange(COMMANDS_DLX, 'topic', { durable: true });

  await channel.assertQueue(COMPANIES_DEAD_QUEUE, { durable: true });
  await channel.bindQueue(COMPANIES_DEAD_QUEUE, DOMAIN_EVENTS_DLX, '#');
}
